### DISCUSSION TITLE
Async function error handling

### DISCUSSION THREAD

**Discussion by kellytk** [tokio-rs/axum#493] [Q&A] [answered]
[The documentation](https://docs.rs/axum/0.3.2/axum/error_handling/index.html#routing-to-fallible-services) states:

    You generally don’t have to think about errors if you’re only using async functions as handlers.

What's the recommended practice if you must?  Consider a simple handler relying upon an Askama template, temporarily with incomplete error handling:

```
async fn handler() -> Html<String> {
    Html(MyAskamaTemplate{}.render().unwrap())
}
```

As this is a necessary resource for the server to operate, this appears to be a valid case of server error and not covered by request-oriented errors discussed in https://docs.rs/axum/0.3.2/axum/error_handling/index.html#axums-error-handling-model.

Is my understanding correct?  If it is, how should async function handlers properly address server errors?  Simply using a `Result<Html<String>, E>`?

**Comment by davidpdrsn** [comment, marked as answer]
You have to handle the error somehow and return an appropriate response to the client. You can return `Result<T, E>` if both types implement `IntoResponse`.

Note this still isn't an "error" in the `tower::Service` sense since a response is still generated and sent back to the client.

**Reply by kellytk** [reply]
And originating in a handler, the Askama error could only exist in response to a request and not at program initialization, correct?

**Reply by davidpdrsn** [reply]
I would guess so but not familiar enough with askama to know for sure.

**Reply by kellytk** [reply]
Is it theoretically possible for an async function handler to generate an error at program initialization (`Router::new().route()`) or could it only be in response to a request?

**Reply by davidpdrsn** [reply]
Overlapping routes will cause panic (see [here](https://docs.rs/axum/0.3.2/axum/routing/struct.Router.html#method.route) for more details). Otherwise not. Any other such things will be detailed in the docs.

**Reply by kellytk** [reply]
Utilizing Result with an error type of (StatusCode, Html<String>) is working well.

I have several handlers that can generate an internal server error and contain code for generating an appropriate response in that case.  Rather than duplicating the error condition response code across multiple handlers, I can move it into a macro.

A potentially better solution would be if handlers that could generate an internal server error, upon that error, passed the request and responsibility to respond, to a dedicated internal server error async function handler.  Is that possible with Axum?

**Reply by davidpdrsn** [reply]
You make your own error enum and implement `IntoResponse` for it like shown here https://docs.rs/axum/0.3.2/axum/response/trait.IntoResponse.html#implementing-intoresponse

**Reply by davidpdrsn** [reply]
And here https://github.com/tokio-rs/axum/blob/main/examples/error-handling-and-dependency-injection/src/main.rs

**Reply by kellytk** [reply]
How can shared state, from the handler or otherwise, find its way into the error enum's `IntoResponse` implementation?

**Reply by davidpdrsn** [reply]
Use an extractor to get the state then pass that state into you `IntoResponse` type

**Reply by kellytk** [reply]
When the handler creates and returns the `IntoResponse` error enum, it should pass the shared state via the error enum variant?

```
enum AppError {
    A,
    B(State),
}
```

**Reply by davidpdrsn** [reply]
Something like this:

```rust
async fn handler(Extension(state): Extension<State>) -> Result<Html<&'static str>, Error> {
    Err(Error::Foo(state))
}

#[derive(Clone)]
struct State {}

enum Error {
    Foo(State),
}

impl IntoResponse for Error {
    type Body = Body;
    type BodyError = <Self::Body as axum::body::HttpBody>::Error;

    fn into_response(self) -> axum::http::Response<Self::Body> {
        todo!()
    }
}
```

**Reply by kellytk** [reply]
I have it compiling with a simple response, such as `Response::new(Body::empty())`.

In the contrived buildable example at https://github.com/kellytk/x-axum-response/blob/main/src/main.rs#L56 you'll see that the error response can be either Xhtml or plain text.  How should that be correctly implemented?

**Reply by davidpdrsn** [reply]
You can build different responses in each branch of the if and if bodies are different types you can use [`box_body`](https://docs.rs/axum/0.3.2/axum/body/fn.box_body.html) to make the types match.
